src/config.rs

```diff
@@ -0,0 +1,1235 @@
+    /// max number of unauthenticated connections
+    pub max_incomplete_connections: u32,
+    /// max number of completed connections from the same user (only enforced on Unix OSs)
+    pub max_connections_per_user: u32,
+    /// max number of service launches in progress at the same time
+    pub max_pending_service_starts: u32,
+    /// max number of names a single connection can own
+    pub max_names_per_connection: u32,
+    /// max number of match rules for a single connection
+    pub max_match_rules_per_connection: u32,
+    /// max number of pending method replies per connection (number of calls-in-progress)
+    pub max_replies_per_connection: u32,
+    /// time until a method call times out
+    pub reply_timeout: Duration,
+}
+
+#[derive(Clone, Debug, Default, Deserialize, PartialEq)]
+#[serde(rename_all = "snake_case")]
+pub enum MessageType {
+    #[default]
+    #[serde(rename = "*")]
+    Any,
+    MethodCall,
+    MethodReturn,
+    Signal,
+    Error,
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub enum Operation {
+    /// rules checked when a new connection to the message bus is established
+    Connect,
+    /// rules checked when a connection attempts to own a well-known bus names
+    Own,
+    /// rules that are checked for each recipient of a message
+    Receive(ReceiveOperation),
+    /// rules that are checked when a connection attempts to send a message
+    Send(SendOperation),
+}
+type OptionalOperation = Option<Operation>;
+impl TryFrom<RuleAttributes> for OptionalOperation {
+    type Error = Error;
+
+    fn try_from(value: RuleAttributes) -> std::result::Result<Self, Self::Error> {
+        let has_connect = false;
+        let has_own = value.own.is_some();
+        let has_send = value.send_broadcast.is_some()
+            || value.send_destination.is_some()
+            || value.send_destination_prefix.is_some()
+            || value.send_error.is_some()
+            || value.send_interface.is_some()
+            || value.send_member.is_some()
+            || value.send_path.is_some()
+            || value.send_requested_reply.is_some()
+            || value.send_type.is_some();
+        let has_receive = value.receive_error.is_some()
+            || value.receive_interface.is_some()
+            || value.receive_member.is_some()
+            || value.receive_path.is_some()
+            || value.receive_sender.is_some()
+            || value.receive_requested_reply.is_some()
+            || value.receive_type.is_some();
+
+        let operations_count: i8 = vec![has_connect, has_own, has_receive, has_send]
+            .into_iter()
+            .map(i8::from)
+            .sum();
+
+        if operations_count > 1 {
+            return Err(Error::msg(format!("do not mix rule attributes for connect, own, receive, and/or send operations in the same rule: {value:?}")));
+        }
+
+        // https://github.com/dbus2/busd/issues/79
+        if value.receive_member.is_some() {
+            eprintln!(
```

Comment:
we've tracing as a dep, please use that for all logging. Then you can remove `warning: ` prefixes as well. :)